File: apps/video-api/src/video_api/capabilities.py
"""Effective deployment capabilities for GET /v1/capabilities.

The create request accepts many knobs whose real behaviour depends on the
server's environment (.env): the TTS engine limits which languages can be
spoken, research and stock media need a configured provider, the "high"
quality profile only adds something when a vision model exists, and the
render engine has a deployment default. This module derives all of that from
one Settings snapshot so clients (Studio in particular) can build their UI
from the actual platform state instead of hardcoding assumptions.

No secrets leave this module: provider *names* are exposed, keys never.
"""
from __future__ import annotations

from video_api.config import Settings
from video_api.languages import SUPPORTED_LANGUAGES
from video_api.schemas import (
    DURATION_MAX_SECONDS,
    DURATION_MIN_SECONDS,
    MAX_BATCH_LANGUAGES,
    PROMPT_MAX_CHARS,
    RESEARCH_SOURCES_DEFAULT,
    RESEARCH_SOURCES_MAX,
    RESEARCH_SOURCES_MIN,
    THEME_MAX_CHARS,
    VISUAL_ASSETS_DEFAULT,
    VISUAL_ASSETS_MAX,
    VISUAL_ASSETS_MIN,
)
from video_api.voices import (
    KOKORO_VOICES,
    QUALITY_PROFILES,
    engine_family_for_profile,
    voices_for_family,
)

RENDER_ENGINES = ("manim", "remotion")

# Languages each TTS family can actually speak. None = every language the API
# accepts (SUPPORTED_LANGUAGES). Kokoro derives from its curated catalog so the
# two can never drift apart; Chatterbox ships an English-only model.
_KOKORO_LANGUAGES = tuple(
    sorted({code for voice in KOKORO_VOICES for code in (voice.languages or ())})
)
_FAMILY_LANGUAGES: dict[str, tuple[str, ...] | None] = {
    "chatterbox": ("en",),
    "kokoro": _KOKORO_LANGUAGES,
}


def languages_for_family(family: str) -> list[str]:
    """Language codes the family can speak, in SUPPORTED_LANGUAGES order."""
    allowed = _FAMILY_LANGUAGES.get(family)
    if allowed is None:
        return list(SUPPORTED_LANGUAGES)
    return [code for code in SUPPORTED_LANGUAGES if code in allowed]


def _provider_feature(provider: str) -> dict:
    normalized = (provider or "").strip().lower()
    available = normalized not in {"", "none"}
    return {"available": available, "provider": normalized if available else None}


def capabilities_payload(settings: Settings) -> dict:
    engine_by_profile = {
        profile: engine_family_for_profile(settings, profile) for profile in QUALITY_PROFILES
    }
    languages_by_profile = {
        profile: languages_for_family(family) for profile, family in engine_by_profile.items()
    }
    voice_selection_by_profile = {
        profile: len(voices_for_family(settings, family)) > 0
        for profile, family in engine_by_profile.items()
    }
    return {
        "engine": engine_by_profile["standard"],
        "engine_by_profile": engine_by_profile,
        "languages": [
            {"code": code, "name": name} for code, name in SUPPORTED_LANGUAGES.items()
        ],
        "languages_by_profile": languages_by_profile,
        "voice_selection_by_profile": voice_selection_by_profile,
        "render_engines": list(RENDER_ENGINES),
        "features": {
            "research": _provider_feature(settings.research_provider),
            "stock_assets": _provider_feature(settings.asset_provider),
            "visual_review": {
                "available": bool(settings.visual_review_model.strip()),
                "provider": None,
            },
        },
        "limits": {
            "prompt_max_chars": PROMPT_MAX_CHARS,
            "theme_max_chars": THEME_MAX_CHARS,
            "max_batch_languages": MAX_BATCH_LANGUAGES,
            "target_duration_seconds": {
                "min": DURATION_MIN_SECONDS,
                "max": DURATION_MAX_SECONDS,
                # The env default is a free integer; keep the advertised default
                # inside the accepted request window.
                "default": min(
                    max(settings.default_target_duration_seconds, DURATION_MIN_SECONDS),
                    DURATION_MAX_SECONDS,
                ),
            },
            "research_max_sources": {
                "min": RESEARCH_SOURCES_MIN,
                "max": RESEARCH_SOURCES_MAX,
                "default": RESEARCH_SOURCES_DEFAULT,
            },
            "visuals_max_assets": {
                "min": VISUAL_ASSETS_MIN,
                "max": VISUAL_ASSETS_MAX,
                "default": VISUAL_ASSETS_DEFAULT,
            },
        },
        "defaults": {
            "production_mode": settings.production_mode,
            "caption_mode": settings.caption_mode,
            "quality_profile": "standard",
            "render_engine": settings.render_engine,
        },
    }

File: apps/studio/src/lib/capabilities.ts
// Normalizes GET /v1/capabilities into the shape the create form consumes.
// The form never hardcodes what the deployment supports: languages come from
// the effective TTS engine per profile, features (research, stock media,
// visual review) from the configured providers, limits from the API contract.
// An older API without the endpoint gets a permissive fallback mirroring the
// server's own built-in contract, so the form still works — it just can't
// grey out what the deployment lacks.

import type {
  CapabilitiesResponse,
  CaptionMode,
  ProductionMode,
  QualityProfile,
  RenderEngine,
  VoiceEngine,
} from "../api/types";

export interface LanguageOption {
  code: string;
  name: string;
}

export interface Range {
  min: number;
  max: number;
  default: number;
}

export interface EffectiveCaps {
  // False while loading or when the API predates /v1/capabilities.
  fromServer: boolean;
  engineByProfile: Record<string, VoiceEngine>;
  // Every language the API accepts, ordered for display (FR labels).
  languages: LanguageOption[];
  languagesByProfile: Record<string, string[]>;
  renderEngines: RenderEngine[];
  research: { available: boolean; provider: string | null };
  stockAssets: { available: boolean; provider: string | null };
  visualReview: { available: boolean };
  limits: {
    promptMaxChars: number;
    themeMaxChars: number;
    maxBatchLanguages: number;
    duration: Range;
    researchMaxSources: Range;
    visualsMaxAssets: Range;
  };
  defaults: {
    productionMode: ProductionMode;
    captionMode: CaptionMode;
    qualityProfile: QualityProfile;
    renderEngine: RenderEngine;
  };
}

// French display names; codes the server advertises that we don't know keep
// the server-provided (English) name.
const FRENCH_LANGUAGE_NAMES: Record<string, string> = {
  en: "Anglais",
  fr: "Français",
  es: "Espagnol",
  it: "Italien",
  pt: "Portugais",
  de: "Allemand",
  nl: "Néerlandais",
  ro: "Roumain",
  pl: "Polonais",
  cs: "Tchèque",
  da: "Danois",
  sv: "Suédois",
  fi: "Finnois",
  el: "Grec",
  hu: "Hongrois",
  mk: "Macédonien",
  ru: "Russe",
  tr: "Turc",
  zh: "Chinois",
  yue: "Cantonais",
  ar: "Arabe",
  he: "Hébreu",
  hi: "Hindi",
  ms: "Malais",
  fa: "Persan",
  sw: "Swahili",
  tl: "Tagalog",
  th: "Thaï",
  vi: "Vietnamien",
};

// Display order: common European targets first, then the rest.
const PREFERRED_ORDER = [
  "fr", "en", "es", "de", "it", "pt", "nl", "ro", "pl", "cs", "da", "sv", "fi", "el", "hu",
];

// Mirror of languages.py SUPPORTED_LANGUAGES, used only when the API has no
// /v1/capabilities endpoint.
const FALLBACK_LANGUAGE_CODES = Object.keys(FRENCH_LANGUAGE_NAMES);

const QUALITY_PROFILES = ["draft", "standard", "high", "final"] as const;

function orderLanguages(options: LanguageOption[]): LanguageOption[] {
  const rank = (code: string) => {
    const index = PREFERRED_ORDER.indexOf(code);
    return index === -1 ? PREFERRED_ORDER.length : index;
  };
  return [...options].sort(
    (a, b) => rank(a.code) - rank(b.code) || a.name.localeCompare(b.name, "fr"),
  );
}

const FALLBACK_CAPS: EffectiveCaps = {
  fromServer: false,
  engineByProfile: {},
  languages: orderLanguages(
    FALLBACK_LANGUAGE_CODES.map((code) => ({ code, name: FRENCH_LANGUAGE_NAMES[code] })),
  ),
  languagesByProfile: Object.fromEntries(
    QUALITY_PROFILES.map((profile) => [profile, FALLBACK_LANGUAGE_CODES]),
  ),
  renderEngines: ["manim", "remotion"],
  // Without capability data we cannot grey anything out: stay permissive and
  // let the server enforce.
  research: { available: true, provider: null },
  stockAssets: { available: true, provider: null },
  visualReview: { available: true },
  limits: {
    promptMaxChars: 4000,
    themeMaxChars: 80,
    maxBatchLanguages: 8,
    duration: { min: 20, max: 900, default: 240 },
    researchMaxSources: { min: 3, max: 20, default: 10 },
    visualsMaxAssets: { min: 0, max: 12, default: 4 },
  },
  defaults: {
    productionMode: "technical",
    captionMode: "off",
    qualityProfile: "standard",
    renderEngine: "manim",
  },
};

export function normalizeCaps(data: CapabilitiesResponse | undefined): EffectiveCaps {
  if (!data) return FALLBACK_CAPS;
  return {
    fromServer: true,
    engineByProfile: data.engine_by_profile,
    languages: orderLanguages(
      data.languages.map((lang) => ({
        code: lang.code,
        name: FRENCH_LANGUAGE_NAMES[lang.code] ?? lang.name,
      })),
    ),
    languagesByProfile: data.languages_by_profile,
    renderEngines: data.render_engines,
    research: data.features.research,
    stockAssets: data.features.stock_assets,
    visualReview: { available: data.features.visual_review.available },
    limits: {
      promptMaxChars: data.limits.prompt_max_chars,
      themeMaxChars: data.limits.theme_max_chars,
      maxBatchLanguages: data.limits.max_batch_languages,
      duration: data.limits.target_duration_seconds,
      researchMaxSources: data.limits.research_max_sources,
      visualsMaxAssets: data.limits.visuals_max_assets,
    },
    defaults: {
      productionMode: data.defaults.production_mode,
      captionMode: data.defaults.caption_mode,
      qualityProfile: data.defaults.quality_profile,
      renderEngine: data.defaults.render_engine,
    },
  };
}

/** Languages allowed for a profile, as displayable options. */
export function allowedLanguages(caps: EffectiveCaps, profile: QualityProfile): LanguageOption[] {
  const codes = caps.languagesByProfile[profile];
  if (!codes) return caps.languages;
  const set = new Set(codes);
  return caps.languages.filter((lang) => set.has(lang.code));
}

/** Human label for a TTS engine family. */
export function engineLabel(engine: VoiceEngine | undefined): string {
  switch (engine) {
    case "kokoro":
      return "Kokoro";
    case "moss":
      return "MOSS";
    case "openai":
      return "OpenAI TTS";
    case "chatterbox":
      return "Chatterbox";
    default:
      return engine ?? "serveur";
  }
}
